code/game/objects/items/weapons/grenades/deconstruction_grenade.dm

```diff
@@ -0,0 +1,288 @@
+		playsound(loc, 'sound/machines/terminal_off.ogg', 60, 0)
+	else
+		return ..()
+
+/obj/item/grenade/deconstruction/attackby(obj/item/O, mob/user, params)
+	if(istype(O, /obj/item/assembly/signaler/anomaly/vortex))
+		if(core)
+			to_chat(user, "<span class='notice'>[src] already has a [O]!</span>")
+			return
+		if(!user.drop_item())
+			to_chat(user, "<span class='warning'>[O] is stuck to your hand!</span>")
+			return
+		to_chat(user, "<span class='notice'>You insert [O] into [src], and [src] starts to warm up.</span>")
+		O.forceMove(src)
+		core = O
+		icon_state = "deconstruction_core"
+		update_icon()
+
+/obj/item/grenade/deconstruction/CtrlClick(mob/living/L)
+	radial_menu(L)
+
+/obj/item/grenade/deconstruction/proc/radial_menu(mob/user)
+	if(!check_menu(user))
+		return
+	var/list/choices = list(
+		DECON_DOOR = image(icon = 'icons/obj/interface.dmi', icon_state = "airlock"),
+		CONSUME_ALL = image(icon = 'icons/obj/interface.dmi', icon_state = "trash"),
+		DECON_WINDOWS = image(icon = 'icons/obj/interface.dmi', icon_state = "grillewindow"),
+		DECON_WALL = image(icon = 'icons/obj/interface.dmi', icon_state = "wall"),
+		RECYCLE = image(icon = 'icons/obj/interface.dmi', icon_state = "recycle")
+	)
+	var/outcome = "on"
+	var/choice = show_radial_menu(user, src, choices, custom_check = CALLBACK(src, .proc/check_menu, user))
+	if(!check_menu(user))
+		return
+	switch(choice)
+		if(DECON_DOOR)
+			if(deconstruct_doors)
+				deconstruct_doors = FALSE
+				outcome = "off"
+			else
+				deconstruct_doors = TRUE
+		if(CONSUME_ALL)
+			if(consume_all)
+				consume_all = FALSE
+				outcome = "off"
+			else
+				consume_all = TRUE
+		if(DECON_WINDOWS)
+			if(deconstruct_windows)
+				deconstruct_windows = FALSE
+				outcome = "off"
+			else
+				deconstruct_windows = TRUE
+		if(DECON_WALL)
+			if(deconstruct_walls)
+				deconstruct_walls = FALSE
+				outcome = "off"
+			else
+				deconstruct_walls = TRUE
+		if(RECYCLE)
+			if(recycle)
+				recycle = FALSE
+				outcome = "off"
+			else
+				recycle = TRUE
+
+	playsound(src, 'sound/effects/pop.ogg', 50, 0)
+	if(choice)
+		to_chat(user, "<span class='notice'>You [choice] to '[outcome]'.</span>")
+
+/obj/item/grenade/deconstruction/proc/check_menu(mob/living/user)
+	if(!istype(user))
+		return FALSE
+	if(user.incapacitated() || !user.Adjacent(src))
+		return FALSE
+	return TRUE
+
+/obj/item/grenade/deconstruction/prime()
+	if(!active) //For if it has been deactivated
+		return
+	unprime()
+	if(get_turf(src) == level_name_to_num(CENTCOMM)) //Safety first.
+		atom_say("Activation denied: Area does not need repossession ")
+		return
+	deconstruct_obj(DECONSTRUCT_ATTEMPTS)
+	addtimer(CALLBACK(src, .proc/reboot), 60 SECONDS)
+	on_cooldown = TRUE
+
+/obj/item/grenade/deconstruction/unprime()
+	..()
+	anchored = FALSE
+	set_light(0)
+
+/obj/item/grenade/deconstruction/proc/deconstruct_obj(loops = 0) //We want structures fully deconstructed, no frames or anything, so multiple go arounds.
```

Comment:
It almost feels like this proc should sleep/use addtimer. Both for effect and for performance hogging. 50 runs doing 3/4 views of range 7 each...
Unsure how the profile looks for this one though. Can you check that?